### PR TITLE
Issue/3248 modules v2 disable import based install

### LINKED ISSUE #3248: Modules V2: don't allow any import statement to trigger a V2 install
- Currently, an import statement in a V1 module can trigger an install of a V2 module
- This is a security issue, as internal V1 modules can be shadowed by public V2 modules via pypi
- This should no longer be allowed.

# BASELINE

- A V2 module install can only be triggered by the module being in `requirements.txt` or `setup.cfg`
- Add `requirements.txt` to projects (also take into account (`-c`) the constraints in project.yml)

### PR DESCRIPTION
**Pull request by sanderr**
# Description

Security fix: disable import based installation for v2 modules.

closes #3248

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] ~~End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/module.py [hunk 1]
```diff
@@ -1155,6 +1155,45 @@ def _remove_line_continuations(self, lines: List[str]) -> List[str]:
...
+    def get_module_v2_requirements(self) -> List[InmantaModuleRequirement]:
+        """
+        Returns all requirements this module like has on v2 modules.
+        """
+        return [
+            InmantaModuleRequirement.parse(ModuleV2Source.get_inmanta_module_name(req))
+            for req in self.get_all_python_requirements_as_list()
+            if req.startswith(ModuleV2.PKG_NAME_PREFIX)
+        ]
+
```
Comment (sanderr) [inline src/inmanta/module.py:1196]: Nothing really new here. I moved some methods up to `ModuleLike` so `Project` has them as well and I added `get_module_v2_requirements`.

File: tests/moduletool/common.py [hunk 2]
```diff
@@ -197,82 +190,3 @@ def delegator(*args, **kw):
...
-            fd.write(new_content_init_cf)
-    with open(config_file, "w") as fh:
-        config.write(fh)
-    if install:
-        ModuleTool().install(editable=editable, path=dest_dir)
-    if publish_index is not None:
-        ModuleTool().build(path=dest_dir, output_dir=publish_index.artifact_dir)
-        publish_index.publish()
-    with open(config_file, "r") as fh:
-        return module.ModuleV2Metadata.parse(fh)
```
Comment (sanderr) [inline tests/moduletool/common.py:278]: I've moved this up a level to `tests/utils.py`.

File: tests/moduletool/test_install.py [hunk 3]
```diff
@@ -462,13 +476,19 @@ def test_project_install_modules_cache_invalid(
...
     if not preinstall_v2:
         # populate project.modules[module_name]
-        module.Project.get().get_module(module_name, install=False, allow_v1=True)
+        module.Project.get().get_module(module_name, allow_v1=True)
 
     os.chdir(module.Project.get().path)
+    # TODO: this fails for the v1 preinstalled test because the v2 modules now get installed before starting the import based
+    # install flow for v1 modules. This means the v1 error is very much an edge case that can only occur when a v2 module is
+    # a transient dependency of a v1 module and this v2 module promotes another v1 to v2. Not sure whether the test should cover
+    # this or whether we should just drop the v1 preinstalled test.
```
Comment (sanderr) [inline tests/moduletool/test_install.py:491]: I'd forgotten about this. I'll still need to fix this but I'll open the PR for review in the meantime anyway because the faster it's merged in the faster the build timeouts will disappear. My fix here should only affect this one test case so you should be able to review all else already.

File: src/inmanta/module.py [hunk 4]
```diff
@@ -1397,19 +1459,80 @@ def require_v2(module_name: str) -> None:
...
+                set_up.remove(module_name)
+            if module_name in done:
                 # some submodules already loaded as v1 => reload
                 imports.extend(done[module_name].values())
                 del done[module_name]
-                del ast_by_top_level_mod[module_name]
+                if module_name in ast_by_top_level_mod:
+                    del ast_by_top_level_mod[module_name]
 
+        def load_module_requirements(module_like: ModuleLike) -> None:
```
Comment (wouterdb) [inline src/inmanta/module.py:1471]: It would be good if this method name made explicit that is about V2 modules

File: src/inmanta/module.py [hunk 5]
```diff
@@ -1366,18 +1415,20 @@ def use_module_cache() -> bool:
 
         if use_module_cache():
             return self.modules[module_name]
-        return self.load_module(module_name, install=install, allow_v1=allow_v1)
+        return self.load_module(module_name, allow_v1=allow_v1, install_v1=install_v1, install_v2=install_v2)
 
     def load_module_recursive(
```
Comment (wouterdb) [inline src/inmanta/module.py:1420]: General questions: from a design perspective, we no longer have loading and install as a single function (i.e. `inmanta project install` does the install, `inmanta compile` the loading). Could we lose some complexity by decoupling these two things in the code flow as well? 

(e.g. if we know no install will happen, reloading of a V1 to V2 can also never happen)

File: tests/moduletool/test_install.py (hunk 3 above)
Comment (wouterdb) [reply tests/moduletool/test_install.py:491]: Perhaps this can be tested by hand once and then left behind?

File: tests/test_module_loader.py [hunk 6]
```diff
@@ -318,3 +332,90 @@ def test_load_module_recursive_complex_module_dependencies(local_module_package_
...
+    allowed.
+    """
+    main_module_name: str = "mymodule"
+    dependency_module_name: str = "minimalv2module"
+    index: PipIndex = PipIndex(os.path.join(str(tmpdir), ".my-index"))
+    libs_dir: str = os.path.join(str(tmpdir), "libs")
+    os.makedirs(libs_dir)
+
+    if v1:
+        shutil.copytree(os.path.join(modules_dir, "minimalv1module"), os.path.join(libs_dir, main_module_name))
```
Comment (wouterdb) [inline tests/test_module_loader.py:384]: Why patch up a module to something which is actually completely different (I think)? 
This is not easy to read (I need to investigate both `minimalv1module` and all this code) and it doesn't very clearly convey what it does.

Is it not easier to use e.g. the `makemodule` helper?

File: src/inmanta/module.py (hunk 5 above)
Comment (sanderr) [reply src/inmanta/module.py:1420]: I think this would be difficult to accomplish as long as we still support v1 modules, that do have import-based installations. I actually started out that way but I ended up keeping it in `load_module_recursive`.

File: src/inmanta/module.py (hunk 4 above)
Comment (arnaudsjs) [inline src/inmanta/module.py:1471]:
```suggestion
        def load_v2_module_requirements(module_like: ModuleLike) -> None:
```

File: src/inmanta/module.py [hunk 7]
```diff
@@ -1386,8 +1437,19 @@ def load_module_recursive(
...
 
-        v2_modules: Set[str] = set()  # Set of modules that should be loaded as a V2 module
+        v2_modules: Set[str] = set()
+        """
+        Set of modules that should be loaded as a V2 module.
+        """
+        set_up: Set[str] = set()
+        """
+        Set of top level modules that have been set up (setup_module()).
+        """
```
Comment (arnaudsjs) [inline src/inmanta/module.py:1447]: Why is the `set_up` set required? I think this needs a bit more documentation.

File: tests/test_module_loader.py (hunk 6 above)
Comment (sanderr) [reply tests/test_module_loader.py:384]: - The idea behind `minimalvxmodule` is that they are minimal so you shouldn't necessarily need to look at them to understand the code.
- I thought it wouldn't be worthwhile to create such a helper for v1 but maybe I was wrong about that one, I'll add it.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/module.py (hunk 4 above)
Comment (sanderr) [reply src/inmanta/module.py]: done

File: tests/test_module_loader.py (hunk 6 above)
Comment (sanderr) [reply tests/test_module_loader.py]: done

File: src/inmanta/module.py (hunk 7 above)
Comment (sanderr) [reply src/inmanta/module.py:1447]: I've expanded the `setup_module` docstring. Does that resolve your concerns?

File: src/inmanta/module.py (hunk 4 above)
Comment (sanderr) [reply src/inmanta/module.py]: I've already renamed this to `load_module_v2_requirements` following a comment by @wouterdb. I assume that's fine as well?

File: tests/moduletool/test_install.py (hunk 3 above)
Comment (sanderr) [reply tests/moduletool/test_install.py]: Now they've started failing again (not sure how that's possible since I don't recall making any changes since they succeeded) so I'll have to look at it again tomorrow in any case.

Comment (sanderr) [reply tests/moduletool/test_install.py]: Stupid mistake, fixed.

**Review by wouterdb — APPROVED** [review]
